### PR TITLE
Add Rudimentary FSDP Support

### PR DESCRIPTION
**Pull request by KyroChi**
This simple PR enables FSDP training using `torch.distributed.fsdp`. I've included some extra logging metrics too. 

I validated the implementation by training a 1.5B GPT2 style model on four 8xH200 nodes. Training took 14.5 hours and reached a validation loss of 2.71 on OWT.

See attached [WandB report](https://api.wandb.ai/links/krchickering-uc-davis/ipxa6jce) for training curves and throughput.

### CODE REVIEW

**Comment by mpdmanash** [comment]
Just curious: Why do we need FSDP when the model already fits on a single GPU and we can do DDP. Yes with FSDP, we can partition the model into multiple GPU freeing up space and thus accomodate more samples (increase batch size per GPU), but that also introduces communication overhead.

Isn't that going to make the training be ultimately be slower than training on DDP with gradient_accumulation?
